lib/js/stacks.js

```diff
@@ -0,0 +1,43 @@
+
+(function () {
+    var application = Stimulus.Application.start();
+
+    window.Stacks = window.Stacks || Object.create(null);
+    Stacks.controllers = Object.create(null);
+
+    function createController(name, data) {
+        var Controller = function () {
+            Stimulus.Controller.apply(this, arguments);
+        };
+        Controller.prototype = Object.create(Stimulus.Controller.prototype, {
+            // This needs to be writable because Stimulus changes the constructor property
+            // on the controller. This shouldn't be a problem, mind you -- because right here
+            // were're defining the property on the *prototype*. But IE11 throws anyway.
+            constructor: { value: Controller, enumerable: false, writable: true }
+        });
+        Object.setPrototypeOf(Controller, Stimulus.Controller);
+
+        var targets;
+        for (var prop in data) if (data.hasOwnProperty(prop)) {
+            if (prop === "targets") {
+                targets = data[prop];
+                Object.defineProperty(Controller, "targets", {
+                    get: function () { return targets; },
+                    enumerable: false
+                });
+            } else {
+                Controller.prototype[prop] = data[prop];
+            }
+        }
+        return Controller;
+    }
+
+    Stacks.addController = function addController(name, data) {
+        if (!/^s-/.test(name)) {
+            throw "Stacks-created Stimulus controller names must start with \"s-\".";
```

Comment:
Is the thought here that Stacks would have its own Stimulus application just for Stacks components and consuming sites would have their own Stimulus application that they'd add their own controllers to?  For example, if a site wanted a component called `so-comment`, they'd create their own application and use something like `StackOverflow.addController('so-comment', { ... })`

Reply:
If we do have multiple applications, is Stimulus smart enough to ignore actions like `data-action="so-comment-flagging-dialog#submit"` in applications that don't have a `s-comment-flagging-dialog` controller or will it spit out a warning/error?  Do we need to namespace `data-so-action` and `data-s-action` in the schemas?

Reply:
I don't have a a settled opinion on how this works eventually.

The reason that this exception is thrown here is that I don't want anyone to rely on a JavaScript API that isn't supported. We have a tendency to move from "don't use this function, it's an internal thing" to "this function cannot be removed or changed ever" pretty quickly.

In step 1, Stacks offers ready-made Stimulus controllers (or, in the *very* first step, one such controller), nothing else.